### ISSUE TITLE
stream GROUP BY excludes null values.

### ISSUE THREAD

**Issue by big-andy-coates** [confluentinc/ksql#5633]
The following QTT test should pass:

```json
    {
      "name": "null values (stream->table)",
      "statements": [
        "CREATE STREAM INPUT (ID INT KEY, VAL INT) WITH (kafka_topic='test_topic', value_format='JSON');",
        "CREATE TABLE OUTPUT AS SELECT ID, COUNT(VAL) AS NOT_NULL, COUNT(VAL) AS COUNT FROM INPUT GROUP BY ID;"
      ],
      "inputs": [
        {"topic": "test_topic", "key": 11, "value": null},
        {"topic": "test_topic", "key": 11, "value": {"VAL": null}},
        {"topic": "test_topic", "key": 11, "value": {"VAL": 10}}
      ],
      "outputs": [
        {"topic": "OUTPUT", "key": 11, "value": {"NOT_NULL": 0, "COUNT": 1}},
        {"topic": "OUTPUT", "key": 11, "value": {"NOT_NULL": 0, "COUNT": 2}},
        {"topic": "OUTPUT", "key": 11, "value": {"NOT_NULL": 1, "COUNT": 3}}
      ]
    }
```

i.e. the first row, which has a known `key` of `11`, should result in an output row, but it doesn't. It's currently filtered out.

**Comment by mikebin** [comment]
This seems expected based on Kafka Streams docs:

> Detailed behavior of KGroupedStream and CogroupedKStream:
> 
> Input records with null keys are ignored.
> When a record key is received for the first time, the initializer is called (and called before the adder).
> Whenever a record with a non-null value is received, the adder is called.

Specifically, the adder function in an aggregation is only called when a non-null message value is received.

**Comment by big-andy-coates** [comment]
@mjsax I know we discussed this behaviour somewhere else recently. This would be a good please to chim in as to _why_ streams does what it does.

> Input records with null keys are ignored.

I'm assuming this is not really meaning it will drop any source record with a `null` key, but rather it will drop any record where the data being grouped-by is `null`. 

For example:

```sql
CREATE STREAM INPUT (K INT KEY, ID INT) WITH (...);
INSERT INTO INPUT VALUES (null, 1);  -- not ignored
INSERT INTO INPUT VALUES (1, null);  -- ignored.

CREATE TABLE OUTPUT AS 
   SELECT COUNT(*) 
   FROM INPUT
   GROUP BY ID;
```

The `OUTPUT` table would contain a single row, from the first insert, with a count of `1`.

NB: Maybe the docs should be updated to make this clear, assuming I'm right.

> Whenever a record with a non-null value is received, the adder is called.

Do we know why this is the case? A `null` value has special meaning for a table, but I don't see why it should for a stream.  If 

In the example in the description of this issue we're grouping by the existing key of the stream, so we're not extracting the new aggregate key from the value, so why do we care if the value is `null`??

**Comment by mjsax** [comment]
> I'm assuming this is not really meaning it will drop any source record with a null key, but rather it will drop any record where the data being grouped-by is null.

Correct: The cited JavaDocs use the term "key" to refer to the grouping attribute, as in Kafka Streams, you always group by the key (ie, if you group by anything else, the "anything else" becomes the key, data is repartitioned, and afterward the aggregation is done per-key using this new key). Ie, `Input record` means an input record for the aggregation operator itself (ie, after a potentially new key was set and data was repartitioned), but not an original source topic record.

> NB: Maybe the docs should be updated to make this clear, assuming I'm right.

I don't see any need to update the JavaDocs. Note, that the cited JavaDocs are from `KGroupedStream` and `CogroupedKStream` that represent the KStream after re-keying and re-partitioning (ie, the grouping attributes are already set as the key).

The root cause for dropping `null` values is left/outer stream-stream joins. For both operators, if we get input `<k,null>` and the `ValueJoiner` would be called, the left/right value of `apply(V1 left, V2 right)` might be `null` -- and this would introduce an ambiguity between a `<k,null>` input and if a key does not match (ie, a left/right outer call of `ValueJoiner#apply`). To avoid the ambiguity we decided that `<k,null>` is considered "malformed" as "there is nothing to be joined" anyway. -- To be consistent across all DSL operators, we apply the same logic (ie, "there is nothing to be aggregated") to aggregations, too.

**Comment by mjsax** [comment]
> i.e. the first row, which has a known key of 11, should result in an output row, but it doesn't. It's currently filtered out.

This is by design. I guess the question is really how kslqDB wants to treat `value=null`. If you want to include them in the computation, you might want to transform them into an object that has all columns set to `NULL` for example. This this case, the `value` processed by KafkaStreams is not `null` any longer and the record would be processed.

What I also don't understand is your expected result: The query is `SELECT ID, COUNT(VAL) AS NOT_NULL, COUNT(VAL) AS COUNT`, ie, you perform the exact some operation `COUNT(VAL)` twice. Thus, why do you expect `NOT_NULL` and `NULL` columns in the output to be different?

Last, why do you want `value=null` and `value = {"VAL": null}` to be processed differently?

**Comment by big-andy-coates** [comment]
> To avoid the ambiguity we decided that <k,null> is considered "malformed" as "there is nothing to be joined" anyway. -- To be consistent across all DSL operators, we apply the same logic (ie, "there is nothing to be aggregated") to aggregations, too.

IMHO, ignoring records where the value is null is wrong. You're making the assumption that people always populate the value, and while this assumption is _generally_ true, it's not _guarenteed_.  Consider a system that wants to notify users of changes. Such a system _could_ be modelled as records being produced to Kafka with the changed object's `ID` stored in the Kafka record's key and nothing else.   Such a stream of data is totally valid.  

Claiming "there is nothing to be joined" and "there is nothing to be aggregated" is wrong to IMHO.  I could join the 'changes' stream containing only the `ID` of the changed entity in its key to a table storing the state of the entity, or I might want to count changes per-entity.

> What I also don't understand is your expected result: The query is SELECT ID, COUNT(VAL) AS NOT_NULL, COUNT(VAL) AS COUNT, ie, you perform the exact some operation COUNT(VAL) twice. Thus, why do you expect NOT_NULL and NULL columns in the output to be different?

Doh! Fixed the example. Now `COUNT(VAL) AS NOT_NULL, COUNT(*) AS COUNT`.

> Last, why do you want value=null and value = {"VAL": null} to be processed differently?

I don't. Both would increment the `COUNT` but not the `NON_NULL` count.

**Comment by mjsax** [comment]
I agree, that the "there is nothing to join/aggregate" arguments is not strong. However:

(1) The main motivation (as said already) was to avoid the ambiguity of `leftJoin` call to `ValueJoiner#apply("something", null)` -- it's not clear if `null` would be non-existing right record or a right record with `null` value. And user might want to compute the join result differently for an inner or left/right join result.

(2) If there is `null` in the value, user could put a "dummy" `mapValue(v -> 1)` in front of the aggregation/join to replace the `null` with something not-null (or maybe use `Void` datatype). Hence, for KafkaStreams users it's easy to work-around it.

(3) RocksDB (the default storage engine) does not allow to store `<key,null>` pairs, because those would be delets (similar to compacted topics). (Maybe not an issue for aggregations, but for stream-stream joins.) Also, KafkaStreams cannot easily replace a `null` with a different sentinel automatically, because any sentinel we pick, might conflict with a proper/allowed value for `value` (thus, we need to fall back to (2) and let users specify a non-conflicting `null` sentinel).

Last, if you want to treat `value=null` the same as `value = {"VAL": null}`, than kslqDB can just apply (2) and transform a `null` into a not-`null` (ie, a "row" object with all columns being `NULL`) and it will resolve the issue. Or is there anything that prevents to do this?

**Comment by PeterLindner** [comment]
>>To avoid the ambiguity we decided that <k,null> is considered "malformed" as "there is nothing to be joined" anyway. -- To be consistent across all DSL operators, we apply the same logic (ie, "there is nothing to be aggregated") to aggregations, too.
>
>IMHO, ignoring records where the value is null is wrong. You're making the assumption that people always populate the value, and while this assumption is generally true, it's not guarenteed.

Hi guys, is this also relevant for tables? I think it's quite common for tables to only have primary keys, especially for m:n relations like:

```sql
CREATE TABLE bookauthors (
  authorid INTEGER PRIMARY KEY, 
  bookid INTEGER PRIMARY KEY
);
```
... which technically has no value

This might be a popular query:
```sql
CREATE TABLE books_written AS
SELECT
COUNT(*)
FROM bookauthors
GROUP BY authorid;
```
I actually don't know how this is currently handled and this could probably be solved with dummy values like described above, but I thought I'd ask anyways

**Comment by mjsax** [comment]
For tables it's slightly different, because for TABLES `<key,null>` from a topic has delete semantics (it's a tombstone) by definition (in contrast to a STREAM, for which `<key,null>` has no special semantics). Hence, if you have schema `<A, B, C>` with `A` mapping to the message key, and `B,C` is mapping to the message value, `<a1,null>` is **_not_** the same as `<a1,<NULL,NULL>>` (for a STREAM, it might be ok to treat/interpret `<a1,null>` as `<a1,<NULL,NULL>>` though -- even if I am personally not 100% if we should/want to do this -- if the input is for example in JSON, and I want to write a `<a1,NULL,NULL>` record into the stream I should rather use `<{A:a1}, {}>` instead of `<{A:a1}, null>` (to me, the later record is malformed, while the former contains a valid JSON object in it's value -- of course, it's a personal preference only and I can see why people might want to be able to process the later record, too.

For the TABLE store itself (and thus for RocksDB) the same issue about "storing `null`" applies though. However, before structured keys, the key content is copied/replicated into the value of the key-value-pair, and the value always stores all attributes -- thus, there is no issue with the table having only key-columns. Ie, in you example, the message would be `<{authorid, bookid},{authorid, bookid}>` (ie, there is never a `null` value).

With structured key work, we want to do two things: allow for composite (multi-attribute keys) _and_ avoid the redundant storage of attributes, ie, all PK columns would only be stored in the key, and all other columns are only stored in the value. Hence, your `bookauthors` table could be represented any longer (as we cannot store the `null` key) when the structured key work is finished (not sure atm, how far along we are already)  -- This is actually a case I never considered so far -- thanks for pointing it out! -- we should find a way to support it even with the new structured key design (otherwise, it would be a regression) \cc @big-andy-coates

**Comment by PeterLindner** [comment]
@mjsax thanks for the detailed explanation

**Comment by big-andy-coates** [comment]
Thanks all.  In conclusion: this is something we should fix in ksqlDB, (in deserializes maybe, when source is a STREAM), if it's something we _Want_ to fix.

**Comment by mjsax** [comment]
> if it's something we _Want_ to fix.

For the aggregation question I agree.

I guess it boils does to the question, if `<key,null>` should be valid input for a STREAM or not (btw, given the structured key work, the same question applies to `<null,value>` assuming that the schema maps at least one attributes to the message key via `KEY`). And this decision might depend on what data users have in practice and what data types are used.

For Avro or JSON for example, if all fields could be optional an "empty" (but non-null) avro-record (or JSON, ie, `{}` or `""`) could represent all attributes being `NULL`, and `value=null` might be consider a corrupted messages if we assume that the value must always be a valid Avro/Json object (and `null` is not really valid Avro or Json). I am not sure how users handle this case in practice? Do we have any data points insights if they would use `null` or `""` or `"{}"` to represent an empty Json? Maybe the Json spec itself could help (I would assume they define what is valid Json and what is not valid Json)?

However, if we consider primitive types in "Kafka format" (not sure for Json primitives: there it might still be the empty string instead of `null` to encode `NULL` ? -- Also not sure how primitive Avro is encoded), `value=null` could mean that the corresponding column is `NULL` though. (For Avro/Json the spec might help again.)

For the other problem about tables with all attributes being in the key, I think there is not much of a judgement call, and we must fix it (it's of course not related to this ticket itself but should have its own ticket).
